002-es-modules.md

```diff
@@ -0,0 +1,511 @@
+  2. Return the result of applying the path search algorithm to `specifier`.
+3. Return the result of applying the module search algorithm to `specifier`.
+
+##### 3.3.1.2. Path Search
+
+1. If it does not throw an error, return the result of applying the file search algorithm to `specifier`.
+2. If it does not throw an error, return the result of applying the directory search algorithm to `specifier`.
+3. Throw an error.
+
+##### 3.3.1.3. File Search
+
+1. If the resource for `specifier` exists, return `specifier`.
+2. For each file extension `[".mjs", ".js", ".json", ".node"]`
+  1. Let `searchable` be a new URL from `specifier`.
+  2. Append the file extension to the pathname of `searchable`.
+  3. If the resource for `searchable` exists, return `searchable`.
+3. Throw an error.
+
+##### 3.3.1.4. Package Main Search
+
+1. If it does not throw an error, return the result of applying the file search algorithm to `specifier`.
+2. If it does not throw an error, return the result of applying the index search algorithm to `specifier`.
+3. Throw an error.
+
+##### 3.3.1.5. Index Search
+
+1. Let `searchable` be a new URL from `specifier`.
+2. If `searchable` does not have a trailing `/` in its pathname append one.
+3. Let `searchable` be the result of applying the URL parser to `./index` with `specifier` as the base URL.
+4. If it does not throw an error, return the result of applying the file search algorithm to `searchable`.
+5. Throw an error.
+
+##### 3.3.1.6. Directory Search
+
+1. Let `dir` be a new URL from `specifier`.
+2. If `dir` does not have a trailing `/` in its pathname append one.
+3. Let `searchable` be the result of applying the URL parser to `./package.json` with `dir` as the base URL.
+5. If the resource for `searchable` exists and it contains a "main" field.
+  1. Let `main` be the result of applying the URL parser to the `main` field with `dir` as the base URL.
+  2. If it does not throw an error, return the result of applying the package main search algorithm to `main`.
+6. If it does not throw an error, return the result of applying the index main search algorithm to `dir`.
+7. Throw an error.
+
+##### 3.3.1.7. Module Search
+
+1. Let `package` be a new URL from the directory containing the importing location.
+2. If `package` does not have a trailing `/` in its pathname append one.
+3. Let `searchable` be the result of applying the URL parser to `./node_modules/${specifier}` with `package` as the base URL.
+4. If it does not throw an error, return the result of applying the file search algorithm to `searchable`.
+5. If it does not throw an error, return the result of applying the directory search algorithm to `searchable`.
+6. Let `parent` be the result of applying the URL parser to `../` with `package` as the base URL.
+7. If it does not throw an error, return the result of applying the module search algorithm to `specifier` with an importing location of `parent`.
```

Comment:
Would fail in step 1 when it tries to grab the directory of `parent` which is `/`. Can clarify in step 1 by saying if the parent is same as importing location it should error.